analyses/independent-samples/independent-samples.R

```diff
@@ -0,0 +1,62 @@
+# independent_samples.R
+
+#' Generate a vector of unique samples
+#' 
+#' The samples from this function will be unique with respect to participants
+#' i.e. only no two samples will come from the same participant. The input list 
+#' should be pre-filtered by `experimental_strategy` and `sample_type`.  
+#' 
+#' 
+#' @param sample_df A data frame of samples, with columns corresponding to those
+#'   in `pbta-histologies.tsv`
+#' @param tumor_types Designates which types of tumors will be included. Options
+#'   are "primary" to include only primary tumors, "prefer_primary" to include
+#'   primary tumors when available, but fall back to other types, or "any" to
+#'   randomly select among all available specimens. As of v5, primary tumors
+#'   are defined as those designated "Initial CNS Tumor" or "Diagnosis" in the
+#'   `tumor_descriptor` field.
+#' @param seed An optional random number seed. 
+#' 
+#' @return a vector of biospecimen ids
+independent_samples <- function(sample_df, 
+                                tumor_types = c("primary", "prefer_primary", "any"), 
+                                seed){
+  tumor_types <- match.arg(tumor_types)
+  if(!missing(seed)){set.seed(seed)}
+  
+  primary_descs <- c("Initial CNS Tumor", "Diagnosis")
+  
+  if(tumor_types == "prefer_primary"){
+    # find cases where non-primary is the only option
+    no_primary <- sample_df %>% 
+      dplyr::group_by(Kids_First_Participant_ID) %>%
+      dplyr::summarize(n_primary = sum(tumor_descriptor %in% primary_descs)) %>%
+      dplyr::filter(n_primary == 0) %>%
+      dplyr::pull(Kids_First_Participant_ID)
+  } else {
+    no_primary <- c()
+  }
+  
+  if(tumor_types %in% c("primary", "prefer_primary")){
+    primary_df <- sample_df %>%
+      dplyr::filter(tumor_descriptor %in% primary_descs)
+    noprimary_df <- sample_df %>%
+      dplyr::filter(Kids_First_Participant_ID %in% no_primary)
+    sample_df <- dplyr::bind_rows(primary_df, noprimary_df)
+  } 
+  
+  # get the samples from the earliest timepoints for each Participant
+  early_samples <- sample_df %>%
+   dplyr::group_by(Kids_First_Participant_ID) %>%
+   dplyr::summarize(age_at_diagnosis_days = min(age_at_diagnosis_days)) %>%
+   dplyr::left_join(sample_df, by = c("Kids_First_Participant_ID", 
+                                      "age_at_diagnosis_days"))
+   
+  # Choose randomly among specimens from the same participant 
+  early_ind <- early_samples %>%
+    dplyr::group_by(Kids_First_Participant_ID) %>%
+    dplyr::summarize(Kids_First_Biospecimen_ID = sample(Kids_First_Biospecimen_ID, 1)) %>%
+    dplyr::pull(Kids_First_Biospecimen_ID)
```

Comment:
Question about this - will this PR result in a sample set for everyone to use or will everyone generate their own sample set, because this randomization will then be different for everyone if the latter is the case.

Reply:
I believe the plan is generating the sample set for everyone in part 2. I have a strong preference for including the sample set for everyone in the data download for single point of truth reasons.

Reply:
Yes, the plan was to have a single sample set in part 2, with a small script that just does that. I just wanted to include the thinking and prelim analysis in this notebook.

Reply:
And when I say single sample set, I was actually thinking there might be up to 4: With and without WXS, and with and without non-primary tumors. Thoughts?

Reply:
I like that idea @jashapiro. Good to start thinking about how to document those in the README.